src/geometry.F90

```diff
@@ -871,81 +872,50 @@ end subroutine distance_to_boundary
 
   subroutine neighbor_lists()
 
-    integer :: i          ! index in cells/surfaces array
-    integer :: j          ! index of surface in cell
-    integer :: i_surface  ! index in count arrays
-    integer, allocatable :: count_positive(:) ! # of cells on positive side
-    integer, allocatable :: count_negative(:) ! # of cells on negative side
-    logical :: positive   ! positive side specified in surface list
-    type(Cell), pointer  :: c
+    integer :: i  ! index in cells/surfaces array
+    integer :: j  ! index in region specification
+    integer :: k  ! surface half-space spec
+    type(VectorInt), allocatable :: neighbor_pos(:)
+    type(VectorInt), allocatable :: neighbor_neg(:)
 
     call write_message("Building neighboring cells lists for each surface...", &
-         &4)
+         4)
 
-    allocate(count_positive(n_surfaces))
-    allocate(count_negative(n_surfaces))
-    count_positive = 0
-    count_negative = 0
+    allocate(neighbor_pos(n_surfaces))
+    allocate(neighbor_neg(n_surfaces))
 
     do i = 1, n_cells
-      c => cells(i)
-
-      ! loop over each region specification
-      do j = 1, size(c%region)
-        i_surface = c % region(j)
-        positive = (i_surface > 0)
-
-        ! Skip any tokens that correspond to operators rather than regions
-        i_surface = abs(i_surface)
-        if (i_surface >= OP_UNION) cycle
-
-        if (positive) then
-          count_positive(i_surface) = count_positive(i_surface) + 1
+      do j = 1, size(cells(i)%region)
+        ! Get token from region specification and skip any tokens that
+        ! correspond to operators rather than regions
+        k = cells(i)%region(j)
+        if (abs(k) >= OP_UNION) cycle
+
+        ! Add this cell ID to neighbor list for k-th surface
+        if (k > 0) then
+          call neighbor_pos(abs(k))%push_back(i)
         else
-          count_negative(i_surface) = count_negative(i_surface) + 1
+          call neighbor_neg(abs(k))%push_back(i)
         end if
       end do
     end do
 
-    ! allocate neighbor lists for each surface
     do i = 1, n_surfaces
-      if (count_positive(i) > 0) then
-        allocate(surfaces(i)%obj%neighbor_pos(count_positive(i)))
-      end if
-      if (count_negative(i) > 0) then
-        allocate(surfaces(i)%obj%neighbor_neg(count_negative(i)))
+      ! Copy positive neighbors to Surface instance
+      j = neighbor_pos(i)%size()
```

Comment:
Is there a reason to assign the size to `j` here rather than evaluate `size()` inside the `if` directly? I personally prefer the latter since it would shorten the code here, and ensure that `j` is only used as the "index in region specification" as is specified in the comment next to its declaration at the top of the function.